src/document/sidebar.ml

```diff
@@ -17,41 +17,14 @@ end = struct
   type t = (Url.Path.t * Inline.one) option Tree.t
 
   open Odoc_model.Sidebar
-  open Odoc_model.Paths.Identifier
 
   let of_lang (dir : PageToc.t) =
-    let rec of_lang ~parent_id ((content, index) : PageToc.t) =
-      let title, parent_id =
-        match index with
-        | Some (index_id, title) -> (Some title, Some (index_id :> Page.t))
-        | None -> (None, (parent_id :> Page.t option))
-      in
-      let entries =
-        List.filter_map
-          (fun id ->
-            match id with
-            | id, PageToc.Entry title ->
-                (* TODO warn on non empty children order if not index page somewhere *)
-                let payload =
-                  let path = Url.Path.from_identifier id in
-                  let content = Comment.link_content title in
-                  Some (path, sidebar_toc_entry id content)
-                in
-                Some { Tree.node = payload; children = [] }
-            | id, PageToc.Dir dir -> Some (of_lang ~parent_id:(Some id) dir))
-          content
-      in
-      let payload =
-        match (title, parent_id) with
-        | None, _ | _, None -> None
-        | Some title, Some parent_id ->
-            let path = Url.Path.from_identifier parent_id in
-            let content = Comment.link_content title in
-            Some (path, sidebar_toc_entry parent_id content)
-      in
-      { Tree.node = payload; children = entries }
-    in
-    of_lang ~parent_id:None dir
+    Tree.map dir ~f:(function
+      | None -> None
+      | Some (parent_id, title) ->
+          let path = Url.Path.from_identifier parent_id in
+          let content = Comment.link_content title in
+          Some (path, sidebar_toc_entry parent_id content))
```

Comment:
@panglesd I simplified this code in regards to what is currently done on master, not on #1220, if this is over-factorized and makes the rebase harder for #1220 later feel free to revert this commit before merging.

Reply:
Looks good!